File: apps/mobile/src/lib/api/notes.ts
import { apiBaseURL } from './config';
import {
  APIRequestError as SharedAPIRequestError,
} from './request-error';
import {
  listNotesResponseSchema,
  noteSchema,
  searchNotesResponseSchema,
} from './schema';
import type { TypedTransport } from './client';
import type { APIErrorResponse } from './request-error';
import type { components } from './generated/schema';

export type Note = {
  author: NoteAuthor;
  body: string;
  categorySlug: string;
  createdAt: number;
  id: string;
  images: NoteImage[];
  placeSlug: string | null;
  title: string;
  updatedAt: number;
  usefulCount: number;
  usefulByCurrentUser: boolean;
};

export type NoteImage = {
  byteSize: number;
  contentType: NoteImageResponse['content_type'];
  createdAt: number;
  height: number;
  id: string;
  position: number;
  updatedAt: number;
  url: string;
  width: number;
};

export type NoteAuthor = {
  displayName: string;
  id: string;
};

export type CreateNoteInput = {
  body: string;
  categorySlug: string;
  clientRequestId: string;
  imageUploadIds?: string[];
  placeSlug?: string | null;
  title: string;
};

export type ListNotesInput = {
  categorySlug?: string;
};

export type SearchNotesInput = {
  categorySlug?: string;
  query: string;
};
export type SearchVersion = components['schemas']['SearchVersion'];
export type RetrievalSource = components['schemas']['RetrievalSource'];

export type SearchNoteResult = {
  note: Note;
  retrievalSource: RetrievalSource;
};

export type SearchNotesResult = {
  results: readonly SearchNoteResult[];
  searchVersion: SearchVersion;
};

type GeneratedSchemas = components['schemas'];
type AuthorSummaryResponse = GeneratedSchemas['AuthorSummary'];
type CreateNoteRequest = GeneratedSchemas['CreateNoteRequest'];
type NoteResponse = GeneratedSchemas['Note'];
type NoteImageResponse = GeneratedSchemas['NoteImage'];

export class APIRequestError extends SharedAPIRequestError {
  constructor(
    status: number,
    body: APIErrorResponse | null = null,
    retryAfter?: number,
  ) {
    super(status, body, retryAfter);
  }
}

export class APIResponseError extends Error {
  constructor() {
    super('api_response_invalid');
  }
}

export type NotesAPI = {
  listNotes(input: ListNotesInput): Promise<Note[]>;
  getNote(id: string): Promise<Note>;
  markNoteUseful(noteID: string): Promise<void>;
  unmarkNoteUseful(noteID: string): Promise<void>;
  searchNotes(input: SearchNotesInput): Promise<SearchNotesResult>;
  createNote(input: CreateNoteInput): Promise<Note>;
};

function rewrapTransportError(error: unknown): never {
  if (error instanceof SharedAPIRequestError) {
    throw new APIRequestError(error.status, error.body, error.retryAfter);
  }
  throw error;
}

export function bindNotesAPI(transport: TypedTransport): NotesAPI {
  return {
    async listNotes(input) {
      try {
        const { data } = await transport.GET('/v1/notes', {
          params: { query: noteListQuery(input) },
        });
        return parseListNotesResponse(data);
      } catch (error) {
        rewrapTransportError(error);
      }
    },

    async getNote(id) {
      try {
        const { data } = await transport.GET('/v1/notes/{note_id}', {
          params: { path: { note_id: id } },
        });
        return parseNoteResponse(data);
      } catch (error) {
        rewrapTransportError(error);
      }
    },

    async markNoteUseful(noteID) {
      try {
        await transport.PUT('/v1/notes/{note_id}/useful', {
          params: { path: { note_id: noteID } },
        });
      } catch (error) {
        rewrapTransportError(error);
      }
    },

    async unmarkNoteUseful(noteID) {
      try {
        await transport.DELETE('/v1/notes/{note_id}/useful', {
          params: { path: { note_id: noteID } },
        });
      } catch (error) {
        rewrapTransportError(error);
      }
    },

    async searchNotes(input) {
      try {
        const { data } = await transport.GET('/v1/search/notes', {
          params: { query: noteSearchQuery(input) },
        });
        return parseSearchNotesResponse(data);
      } catch (error) {
        rewrapTransportError(error);
      }
    },

    async createNote(input) {
      const request: CreateNoteRequest = {
        body: input.body,
        category_slug: input.categorySlug,
        client_request_id: input.clientRequestId,
        image_upload_ids: input.imageUploadIds,
        place_slug: input.placeSlug ?? null,
        title: input.title,
      };
      try {
        const { data } = await transport.POST('/v1/notes', { body: request });
        return parseNoteResponse(data);
      } catch (error) {
        rewrapTransportError(error);
      }
    },
  };
}

function noteListQuery(input: ListNotesInput): {
  category_slug?: string;
} {
  if (input.categorySlug === undefined) {
    return {};
  }
  return { category_slug: input.categorySlug };
}

function noteSearchQuery(input: SearchNotesInput): {
  category_slug?: string;
  q: string;
} {
  if (input.categorySlug === undefined) {
    return { q: input.query };
  }
  return { category_slug: input.categorySlug, q: input.query };
}

function parseListNotesResponse(value: unknown): Note[] {
  const listNotesResponse = listNotesResponseSchema.safeParse(value);
  if (!listNotesResponse.success) {
    throw new APIResponseError();
  }

  return listNotesResponse.data.notes.map(mapNoteResponse);
}
export function parseSearchNotesResponse(value: unknown): SearchNotesResult {
  const searchNotesResponse = searchNotesResponseSchema.safeParse(value);
  if (!searchNotesResponse.success) {
    throw new APIResponseError();
  }

  return {
    results: searchNotesResponse.data.results.map((result) => ({
      note: mapNoteResponse(result.note),
      retrievalSource: result.retrieval_source,
    })),
    searchVersion: searchNotesResponse.data.search_version,
  };
}

export function parseNoteResponse(value: unknown): Note {
  const noteResponse = noteSchema.safeParse(value);
  if (!noteResponse.success) {
    throw new APIResponseError();
  }

  return mapNoteResponse(noteResponse.data);
}

export function mapNoteResponse(value: NoteResponse): Note {
  return {
    author: parseAuthorSummary(value.author),
    body: value.body,
    categorySlug: value.category_slug,
    createdAt: value.created_at,
    id: value.id,
    images: value.images.map(parseNoteImage),
    placeSlug: value.place_slug,
    title: value.title,
    updatedAt: value.updated_at,
    usefulCount: value.useful_count,
    usefulByCurrentUser: value.useful_by_current_user,
  };
}

function parseAuthorSummary(value: AuthorSummaryResponse): NoteAuthor {
  return {
    displayName: value.display_name,
    id: value.id,
  };
}

function parseNoteImage(value: NoteImageResponse): NoteImage {
  return {
    byteSize: value.byte_size,
    contentType: value.content_type,
    createdAt: value.created_at,
    height: value.height,
    id: value.id,
    position: value.position,
    updatedAt: value.updated_at,
    url: resolveNoteImageURL(value.url),
    width: value.width,
  };
}

function resolveNoteImageURL(value: string): string {
  try {
    if (isAbsoluteURL(value)) {
      return value;
    }
    return new URL(value, apiBaseURL()).toString();
  } catch {
    throw new APIResponseError();
  }
}

function isAbsoluteURL(value: string): boolean {
  try {
    new URL(value);
    return true;
  } catch {
    return false;
  }
}

File: services/api/internal/httpapi/notes.go
package httpapi

import (
	"context"
	"errors"
	"net/http"

	"github.com/tprei/sdds/services/api/internal/note"
	"github.com/tprei/sdds/services/api/internal/openapi"
	"github.com/tprei/sdds/services/api/internal/user"
)

const recentNotesLimit = note.ListDefaultLimit
const searchNotesLimit = note.SearchDefaultLimit
const maxCreateNoteRequestBytes int64 = 32 * 1024
const maxUpdateNoteRequestBytes int64 = 32 * 1024

func (handler server) ListNotes(w http.ResponseWriter, r *http.Request, params openapi.ListNotesParams) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}
	input := listNotesInput(params)
	input.ViewerUserID = current.User.ID
	if problems := note.ValidateListInput(input); len(problems) > 0 {
		writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidNote, problems))
		return
	}
	if problems, err := handler.validateCategoryFilter(r.Context(), input.CategorySlug); err != nil {
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	} else if len(problems) > 0 {
		writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidNote, problems))
		return
	}

	notes, err := handler.notes.noteStore.ListRecentNotes(r.Context(), input)
	if err != nil {
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	writeJSON(w, http.StatusOK, newListNotesResponse(notes))
}

func (handler server) GetNote(w http.ResponseWriter, r *http.Request, noteID string) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}
	found, err := handler.notes.noteStore.FindNote(r.Context(), noteID, current.User.ID)
	if errors.Is(err, note.ErrNoteNotFound) {
		writeError(w, http.StatusNotFound, openapi.ErrorResponse{Code: openapi.ErrorCodeNotFound})
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	writeJSON(w, http.StatusOK, newNoteResponse(found))
}

func (handler server) UpdateNote(w http.ResponseWriter, r *http.Request, noteID string) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}
	var request openapi.UpdateNoteRequest
	if !decodeJSONRequest(w, r, maxUpdateNoteRequestBytes, &request) {
		return
	}
	updated, err := handler.notes.noteEditor.Edit(r.Context(), updateNoteInput(request, noteID, current.User.ID))
	if err != nil {
		writeNoteEditError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, newNoteResponse(updated))
}

func (handler server) DeleteNote(w http.ResponseWriter, r *http.Request, noteID string) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}
	if err := handler.notes.noteEditor.Delete(r.Context(), noteID, current.User.ID); err != nil {
		writeNoteEditError(w, err)
		return
	}
	noContent(w, r)
}

func updateNoteInput(request openapi.UpdateNoteRequest, noteID string, userID user.UserID) note.EditInput {
	input := note.EditInput{NoteID: noteID, UserID: userID, Title: request.Title, Body: request.Body}
	if request.CategorySlug != nil {
		slug := note.CategorySlug(*request.CategorySlug)
		input.CategorySlug = &slug
	}
	return input
}

func writeNoteEditError(w http.ResponseWriter, err error) {
	var editValidationErr *note.EditValidationError
	if errors.As(err, &editValidationErr) {
		writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidNote, editValidationErr.ValidationProblems()))
		return
	}
	var catalogValidationErr *note.CatalogValidationError
	if errors.As(err, &catalogValidationErr) {
		writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidNote, catalogValidationErr.ValidationProblems()))
		return
	}
	switch {
	case errors.Is(err, note.ErrNoteNotFound):
		writeError(w, http.StatusNotFound, openapi.ErrorResponse{Code: openapi.ErrorCodeNotFound})
	case errors.Is(err, note.ErrNoteForbidden):
		writeError(w, http.StatusForbidden, openapi.ErrorResponse{Code: openapi.ErrorCodeForbidden})
	case errors.Is(err, note.ErrEmbeddingUnavailable):
		writeError(w, http.StatusServiceUnavailable, openapi.ErrorResponse{Code: openapi.ErrorCodeEmbeddingUnavailable})
	default:
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
	}
}

func (handler server) MarkNoteUseful(w http.ResponseWriter, r *http.Request, noteID string) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}

	if _, err := handler.notes.noteStore.FindNote(r.Context(), noteID, current.User.ID); err != nil {
		if errors.Is(err, note.ErrNoteNotFound) {
			writeError(w, http.StatusNotFound, openapi.ErrorResponse{Code: openapi.ErrorCodeNotFound})
			return
		}
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	if err := handler.notes.usefulStore.MarkUseful(r.Context(), note.MarkUsefulInput{NoteID: noteID, UserID: current.User.ID}); err != nil {
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	noContent(w, r)
}

func (handler server) UnmarkNoteUseful(w http.ResponseWriter, r *http.Request, noteID string) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}

	if _, err := handler.notes.noteStore.FindNote(r.Context(), noteID, current.User.ID); err != nil {
		if errors.Is(err, note.ErrNoteNotFound) {
			writeError(w, http.StatusNotFound, openapi.ErrorResponse{Code: openapi.ErrorCodeNotFound})
			return
		}
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	if err := handler.notes.usefulStore.UnmarkUseful(r.Context(), note.UnmarkUsefulInput{NoteID: noteID, UserID: current.User.ID}); err != nil {
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	noContent(w, r)
}

func (handler server) CreateNote(w http.ResponseWriter, r *http.Request) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}

	var request openapi.CreateNoteRequest
	if !decodeJSONRequest(w, r, maxCreateNoteRequestBytes, &request) {
		return
	}

	input := createNoteInput(request, current.User.ID)
	if problems := note.ValidateCreateInput(input); len(problems) > 0 {
		status := http.StatusBadRequest
		code := openapi.ErrorCodeInvalidNote
		for _, problem := range problems {
			if problem.Field == "image_upload_ids" && problem.Message == "too_long" {
				status = http.StatusConflict
				code = openapi.ErrorCodeTooManyImages
				break
			}
		}
		writeError(w, status, validationErrorResponse(code, problems))
		return
	}

	created, err := handler.notes.notePublisher.Publish(r.Context(), input)
	if err != nil {
		var catalogValidationErr *note.CatalogValidationError
		if errors.As(err, &catalogValidationErr) {
			writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidNote, catalogValidationErr.ValidationProblems()))
			return
		}
		switch {
		case errors.Is(err, note.ErrIdempotencyConflict):
			writeError(w, http.StatusConflict, openapi.ErrorResponse{Code: openapi.ErrorCodeIdempotencyConflict})
		case errors.Is(err, note.ErrImageUploadExpired):
			writeError(w, http.StatusConflict, openapi.ErrorResponse{Code: openapi.ErrorCodeUploadExpired})
		case errors.Is(err, note.ErrImageUploadUnavailable):
			writeError(w, http.StatusConflict, validationErrorResponse(openapi.ErrorCodeInvalidNote, []note.ValidationProblem{{Field: "image_upload_ids", Message: "invalid"}}))
		case errors.Is(err, note.ErrCategoryNotFound):
			writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidNote, []note.ValidationProblem{{Field: "category_slug", Message: "unknown"}}))
		case errors.Is(err, note.ErrEmbeddingUnavailable):
			writeError(w, http.StatusServiceUnavailable, openapi.ErrorResponse{Code: openapi.ErrorCodeEmbeddingUnavailable})
		default:
			writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		}
		return
	}

	writeJSON(w, http.StatusCreated, newNoteResponse(created))
}

func (handler server) SearchNotes(w http.ResponseWriter, r *http.Request, params openapi.SearchNotesParams) {
	current, ok := currentSessionFromContext(r.Context())
	if !ok {
		writeUnauthenticated(w)
		return
	}
	input := searchNoteInput(params)
	input.ViewerUserID = current.User.ID
	if problems := note.ValidateSearchInput(input); len(problems) > 0 {
		writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidSearch, problems))
		return
	}
	if problems, err := handler.validateCategoryFilter(r.Context(), input.CategorySlug); err != nil {
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	} else if len(problems) > 0 {
		writeError(w, http.StatusBadRequest, validationErrorResponse(openapi.ErrorCodeInvalidSearch, problems))
		return
	}

	results, err := handler.notes.noteSearcher.Search(r.Context(), input)
	if err != nil {
		if errors.Is(err, note.ErrSearchUnavailable) {
			writeError(w, http.StatusServiceUnavailable, openapi.ErrorResponse{Code: openapi.ErrorCodeEmbeddingUnavailable})
			return
		}
		writeError(w, http.StatusInternalServerError, openapi.ErrorResponse{Code: openapi.ErrorCodeInternal})
		return
	}

	writeJSON(w, http.StatusOK, newSearchNotesResponse(results))
}

func validationErrorResponse(code openapi.ErrorCode, problems []note.ValidationProblem) openapi.ErrorResponse {
	fields := make([]openapi.ValidationProblem, 0, len(problems))
	for _, problem := range problems {
		fields = append(fields, openapi.ValidationProblem{
			Field: openapi.ValidationField(problem.Field),
			Code:  openapi.ValidationProblemCode(problem.Message),
		})
	}
	return openapi.ErrorResponse{Code: code, Fields: &fields}
}

func listNotesInput(params openapi.ListNotesParams) note.ListInput {
	categorySlug := note.CategorySlug("")
	if params.CategorySlug != nil {
		categorySlug = note.CategorySlug(*params.CategorySlug)
	}
	return note.NormalizeListInput(note.ListInput{
		CategorySlug: categorySlug,
		Limit:        recentNotesLimit,
	})
}

func createNoteInput(request openapi.CreateNoteRequest, userID user.UserID) note.CreateInput {
	imageUploadIDs := []string(nil)
	if request.ImageUploadIds != nil {
		imageUploadIDs = *request.ImageUploadIds
	}
	return note.NormalizeCreateInput(note.CreateInput{
		UserID:          userID,
		Title:           request.Title,
		Body:            request.Body,
		CategorySlug:    note.CategorySlug(request.CategorySlug),
		ClientRequestID: request.ClientRequestId,
		ImageUploadIDs:  imageUploadIDs,
	})
}

func (handler server) validateCategoryFilter(ctx context.Context, slug note.CategorySlug) ([]note.ValidationProblem, error) {
	if slug == "" {
		return nil, nil
	}

	if _, err := handler.notes.categoryCatalog.FindActiveCategory(ctx, slug); errors.Is(err, note.ErrCategoryNotFound) {
		return []note.ValidationProblem{{Field: "category_slug", Message: "unknown"}}, nil
	} else if err != nil {
		return nil, err
	}

	return nil, nil
}

func searchNoteInput(params openapi.SearchNotesParams) note.SearchInput {
	query := ""
	if params.Q != nil {
		query = *params.Q
	}
	categorySlug := note.CategorySlug("")
	if params.CategorySlug != nil {
		categorySlug = note.CategorySlug(*params.CategorySlug)
	}
	return note.NormalizeSearchInput(note.SearchInput{
		CategorySlug: categorySlug,
		Query:        query,
		Limit:        searchNotesLimit,
	})
}

func newListNotesResponse(notes []note.Note) openapi.ListNotesResponse {
	response := openapi.ListNotesResponse{Notes: make([]openapi.Note, 0, len(notes))}
	for _, found := range notes {
		response.Notes = append(response.Notes, newNoteResponse(found))
	}
	return response
}
func newSearchNotesResponse(results []note.SearchResult) openapi.SearchNotesResponse {
	response := openapi.SearchNotesResponse{
		SearchVersion: openapi.SearchVersion(note.CurrentSearchVersion),
		Results:       make([]openapi.SearchNoteResult, 0, len(results)),
	}
	for _, result := range results {
		response.Results = append(response.Results, openapi.SearchNoteResult{
			Note:            newNoteResponse(result.Note),
			RetrievalSource: openapi.RetrievalSource(result.RetrievalSource),
		})
	}
	return response
}

func newNoteResponse(found note.Note) openapi.Note {
	images := make([]openapi.NoteImage, 0, len(found.Images))
	for _, image := range found.Images {
		images = append(images, openapi.NoteImage{
			Id:          image.ID,
			Url:         "/v1/media/images/" + image.ID,
			ContentType: openapi.NoteImageContentType(image.ContentType),
			ByteSize:    image.ByteSize,
			Width:       int32(image.Width),
			Height:      int32(image.Height),
			Position:    int32(image.Position),
			CreatedAt:   image.CreatedAt.UTC().UnixMilli(),
			UpdatedAt:   image.UpdatedAt.UTC().UnixMilli(),
		})
	}

	return openapi.Note{
		Id:           found.ID,
		Title:        found.Title,
		Body:         found.Body,
		CategorySlug: openapi.CategorySlug(found.CategorySlug),
		Author: openapi.AuthorSummary{
			Id:          string(found.Author.ID),
			DisplayName: found.Author.DisplayName,
		},
		Images:              images,
		UsefulCount:         found.UsefulCount,
		UsefulByCurrentUser: found.UsefulByCurrentUser,
		CreatedAt:           found.CreatedAt.UTC().UnixMilli(),
		UpdatedAt:           found.UpdatedAt.UTC().UnixMilli(),
	}
}
